Add bazel actions workflow

rsemenov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61040
* **#61039**

- Added a new template for bazel GH Actions workflow
- Simplified the workflow based on @malfet's suggestion by combining build and test jobs into one as we only run a small subset of tests for bazel
- Tested the run to make sure it succeeds
- Build step takes 4 minutes, test step takes 7 minutes

The downside of this approach is that I duplicated some of the jobs in a new template file.  Alternative solution would be to use something like this https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance, however, that is better to be done in a separate PR as linux and windows workflows would need to be changed. Another solution is to use a bunch of if else statements in a linux workflow template to accommodate bazel build as part of it, but this seems not as clean as template inheritance with jinja.  

Here is a link to the latest bazel run with this change https://github.com/pytorch/pytorch/actions/runs/1004656584

Differential Revision: [D29562260](https://our.internmc.facebook.com/intern/diff/D29562260)

rsemenov committed:
Add bazel actions workflow

[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"




[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"


- Added a new template for bazel GH Actions workflow
- Simplified the workflow based on @malfet's suggestion by combining build and test jobs into one as we only run a small subset of tests for bazel
- Tested the run to make sure it succeeds
- Build step takes 4 minutes, test step takes 7 minutes

The downside of this approach is that I duplicated some of the jobs in a new template file.  Alternative solution would be to use something like this https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance, however, that is better to be done in a separate PR as linux and windows workflows would need to be changed. Another solution is to use a bunch of if else statements in a linux workflow template to accommodate bazel build as part of it, but this seems not as clean as template inheritance with jinja.  


[ghstack-poisoned]

rsemenov committed:
Update .github/templates/bazel_ci_workflow.yml.j2

Co-authored-by: Rong Rong <rongr@fb.com>

malfet commented on .github/templates/bazel_ci_workflow.yml.j2:
```diff
@@ -0,0 +1,287 @@
+      - name: Check if image should be built
+        id: check
+        env:
+          DOCKER_TAG: ${{ steps.calculate-tag.outputs.docker_tag }}
+          BASE_REVISION: ${{ github.event.pull_request.base.sha || github.sha }}
+        run: |
+          set -x
+          # Check if image already exists, if it does then skip building it
+          if docker manifest inspect "${DOCKER_IMAGE_BASE}:${DOCKER_TAG}"; then
+            exit 0
+          fi
+          if [[ "$BASE_REVISION" = "$(git rev-parse HEAD)" ]]; then
+            # if we're on the base branch then use the parent commit
+            MERGE_BASE=$(git rev-parse HEAD~)
+          else
+            # otherwise we're on a PR, so use the most recent base commit
+            MERGE_BASE=$(git merge-base HEAD "$BASE_REVISION")
+          fi
+          # Covers the case where a previous tag doesn't exist for the tree
+          # this is only really applicable on trees that don't have `.circleci/docker` at its merge base, i.e. nightly
+          if ! git rev-parse "$MERGE_BASE:.circleci/docker"; then
+            echo "Directory '.circleci/docker' not found in commit $MERGE_BASE, you should probably rebase onto a more recent commit"
+            exit 1
+          fi
+          PREVIOUS_DOCKER_TAG=$(git rev-parse "$MERGE_BASE:.circleci/docker")
+          # If no image exists but the hash is the same as the previous hash then we should error out here
+          if [[ "${PREVIOUS_DOCKER_TAG}" = "${DOCKER_TAG}" ]]; then
+            echo "ERROR: Something has gone wrong and the previous image isn't available for the merge-base of your branch"
+            echo "       contact the PyTorch team to restore the original images"
+            exit 1
+          fi
+          echo ::set-output name=rebuild::yes
+      - name: Build and push docker image
+        if: steps.check.outputs.rebuild
+        env:
+          DOCKER_TAG: ${{ steps.calculate-tag.outputs.docker_tag }}
+          DOCKER_SKIP_S3_UPLOAD: 1
+        run: |
+          export IMAGE_NAME=${DOCKER_IMAGE_BASE#308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/}
+          cd .circleci/docker && ./build_docker.sh
```
This section seems to be word-by-word copy of one from linux_ci_workflow.yml.j2 , see small snippet below 
https://github.com/pytorch/pytorch/blob/864dcbb2cc65352a9ba260c3cf05ee30f4bcc23d/.github/templates/linux_ci_workflow.yml.j2#L36-L40

I wonder if we can modify our template generating system to merge those two somehow

rsemenov replied:
I mentioned this in the PR description, there is inheritance concept with jinya that I think we should start using https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance . I will try it and see how much change that would be to fit in this PR.

rsemenov replied:
Updated to use inheritance. Going forward we should start using [macros](https://jinja.palletsprojects.com/en/3.0.x/templates/#macros). They act like functions and are great for reusability. With macros we potentially could reuse `render_test_results` with passing in some parameters. But this introduces unnecessary complexity at the moment.

rsemenov committed:
Update on "Add bazel actions workflow"


- Added a new template for bazel GH Actions workflow
- Simplified the workflow based on @malfet's suggestion by combining build and test jobs into one as we only run a small subset of tests for bazel
- Tested the run to make sure it succeeds
- Build step takes 4 minutes, test step takes 7 minutes

The downside of this approach is that I duplicated some of the jobs in a new template file.  Alternative solution would be to use something like this https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance, however, that is better to be done in a separate PR as linux and windows workflows would need to be changed. Another solution is to use a bunch of if else statements in a linux workflow template to accommodate bazel build as part of it, but this seems not as clean as template inheritance with jinja.  

Here is a link to the latest bazel run with this change https://github.com/pytorch/pytorch/actions/runs/1004656584

[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"


- Added a new template for bazel GH Actions workflow
- Simplified the workflow based on @malfet's suggestion by combining build and test jobs into one as we only run a small subset of tests for bazel
- Tested the run to make sure it succeeds
- Build step takes 4 minutes, test step takes 7 minutes

The downside of this approach is that I duplicated some of the jobs in a new template file.  Alternative solution would be to use something like this https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance, however, that is better to be done in a separate PR as linux and windows workflows would need to be changed. Another solution is to use a bunch of if else statements in a linux workflow template to accommodate bazel build as part of it, but this seems not as clean as template inheritance with jinja.  

Here is a link to the latest bazel run with this change https://github.com/pytorch/pytorch/actions/runs/1004656584

[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"


- Added a new template for bazel GH Actions workflow
- Simplified the workflow based on @malfet's suggestion by combining build and test jobs into one as we only run a small subset of tests for bazel
- Tested the run to make sure it succeeds
- Build step takes 4 minutes, test step takes 7 minutes

The downside of this approach is that I duplicated some of the jobs in a new template file.  Alternative solution would be to use something like this https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance, however, that is better to be done in a separate PR as linux and windows workflows would need to be changed. Another solution is to use a bunch of if else statements in a linux workflow template to accommodate bazel build as part of it, but this seems not as clean as template inheritance with jinja.  

Here is a link to the latest bazel run with this change https://github.com/pytorch/pytorch/actions/runs/1004656584

[ghstack-poisoned]

rsemenov committed:
Update on "Add bazel actions workflow"


- Added a new template for bazel GH Actions workflow
- Simplified the workflow based on @malfet's suggestion by combining build and test jobs into one as we only run a small subset of tests for bazel
- Tested the run to make sure it succeeds
- Build step takes 4 minutes, test step takes 7 minutes

The downside of this approach is that I duplicated some of the jobs in a new template file.  Alternative solution would be to use something like this https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance, however, that is better to be done in a separate PR as linux and windows workflows would need to be changed. Another solution is to use a bunch of if else statements in a linux workflow template to accommodate bazel build as part of it, but this seems not as clean as template inheritance with jinja.  

Here is a link to the latest bazel run with this change https://github.com/pytorch/pytorch/actions/runs/1004656584

[ghstack-poisoned]

rsemenov commented:
@rsemenov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562260).
